File: C++/Medium/Accepted/can-i-win_1.cpp
class Solution {
public:
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
        
        int sum = maxChoosableInteger * (maxChoosableInteger + 1) / 2;
        if (sum < desiredTotal) return false;
        if (desiredTotal <= 0) return true;
        unordered_map<int, bool> cache;
        return dfs(maxChoosableInteger, desiredTotal, 0, cache);
    }
    bool dfs(int maxChoosableInteger, int desiredTotal, int state, unordered_map<int, bool>& cache) {
        if (cache.count(state)) return cache[state];
        for (int i = 0; i < maxChoosableInteger; ++i) {
            int cur = 1 << i;
            if ((state & cur) == 0) {
                if (desiredTotal <= i + 1 || !dfs(maxChoosableInteger, desiredTotal - (i + 1), state | cur, cache)) {
                    cache[state] = true;
                    return true;
                }
            }
        }
        cache[state] = false;
        return false;
    }
};

File: Java/Medium/Accepted/can-i-win_1.java
class Solution {
    public boolean canIWin(int maxChoosableInteger, int desiredTotal) {
        
        if (desiredTotal <= maxChoosableInteger) {
            return true;
        }
        if ((1 + maxChoosableInteger) * maxChoosableInteger / 2 < desiredTotal) {
            return false;
        }
        Map<Integer, Boolean> map = new HashMap<>();
        return canIWin(maxChoosableInteger, desiredTotal, 0, map);
    }
    private boolean canIWin(int maxChoosableInteger, int desiredTotal, int used, Map<Integer, Boolean> map) {
        if (map.containsKey(used)) {
            return map.get(used);
        }
        for (int i = 1; i <= maxChoosableInteger; i++) {
            int cur = 1 << i;
            if ((used & cur) == 0) {
                if (i >= desiredTotal || !canIWin(maxChoosableInteger, desiredTotal - i, used | cur, map)) {
                    map.put(used, true);
                    return true;
                }
            }
        }
        map.put(used, false);
        return false;
    }
}